Stage 8, sending a message back

4meech opened the pull request:
Оно упало при сообщениии "It's alive!" и каждый раз пыталось прожевать строку с символом , поэтому пришлось нагуглить URLEncoder.

Мне не до конца понятен момент с прибавлением единицы к id апдейта из видосов прошлого  — можешь объяснить? Таким образом мы в цикле перебираем все пришедшие апдейты начиная с самых старых? Сначала в JSON'е ищется первое вхождение update_id -> кастится в целое число и прибавляется 1 для поиска следующего вхождения update_id и передачи его как offset?

4meech committed:
Stage-VIII-1: reply (sending message) method implemented

danelyan commented on src/main/kotlin/Telegram.kt:
```diff
@@ -1,33 +1,38 @@
-        if (startUpdateId == -1 || endUpdateId == -1) continue
-        val updateIdString: String = updates.substring(startUpdateId + 11, endUpdateId)
-        updateId = updateIdString.toInt() + 1
+        val updateIdRegex = "\"update_id\":(.+?),".toRegex()
```
давай не будем создавать объекты Regex каждую итерацию цикла.
Лучше вынести создание всех экземпляров до начала while.
И в целом, лучше стараться не создавать лишних переменных в циклах.

danelyan commented on src/main/kotlin/TelegramBotService.kt:
```diff
@@ -0,0 +1,28 @@
+        return responseUpdate.body()
+    }
+
+    fun sendMessage(botToken: String, chatId: String, message: String) {
```
токен мы можем передать сразу в конструктор TelegramBotService, и не пробрасывать в каждый метод, так как он особо не меняется

danelyan commented on src/main/kotlin/TelegramBotService.kt:
```diff
@@ -0,0 +1,28 @@
+
+class TelegramBotService {
+    fun getUpdates(botToken: String, updateId: Int?): String {
+        val urlGetUpdates = "https://api.telegram.org/bot$botToken/getUpdates?offset=$updateId"
```
Повторяющуюся часть урла можно вынести в константы

danelyan commented on src/main/kotlin/Telegram.kt:
```diff
@@ -1,33 +1,38 @@
+        val updateIdRegex = "\"update_id\":(.+?),".toRegex()
+        val updateIdMatchResult: MatchResult? = updateIdRegex.find(updates)
+        val idGroups = updateIdMatchResult?.groups
+        updateId = idGroups?.get(1)?.value?.toInt()?.plus(1)
```
> Мне не до конца понятен момент с прибавлением единицы к id апдейта из видосов прошлого — можешь объяснить? Таким образом мы в цикле перебираем все пришедшие апдейты начиная с самых старых? Сначала в JSON'е ищется первое вхождение update_id -> кастится в целое число и прибавляется 1 для поиска следующего вхождения update_id и передачи его как offset?

Согласно документации (https://core.telegram.org/bots/api#getupdates):

> Identifier of the first update to be returned. Must be greater by one than the highest among the identifiers of previously received updates.

То есть, мы передаем в параметры запроса getUpdates() параметр offset. И он должен быть на еденицу выше иденификатора ранее полученных обновлений

danelyan commented on src/main/kotlin/TelegramBotService.kt:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    fun sendMessage(botToken: String, chatId: String, message: String) {
+        val encodedMessage = URLEncoder.encode(message, StandardCharsets.UTF_8.toString())
```
> Оно упало при сообщениии "It's alive!" и каждый раз пыталось прожевать строку с символом , поэтому пришлось нагуглить URLEncoder.

Да, потому что в этой фразе есть спецсимволы (ковычка и воскл. знак) которые нельзя передавать в качестве параметров в GET запрос. Их нужно энкодить - что ты и сделал.

Круто что разобрался и сам нашел решение! Это, по сути, чуть ли не основной навык программиста сейчас

4meech committed:
Stage-VIII-2:
repetitive URL-part is a constant now;
botToken moved to TelegramBotService class constructor;
creation of Regex instances is moved out of a while loop;
getUpdates() and sendMessage()  methods are now using one common instance of HttpClient declared as private property "httpClient"

4meech commented:
requestUpdate по аналогии с httpClient тоже хочется вынести в поле класса. При помощи какого-нибудь string builder, передавая разные части строки? Или не стоит так делать?

danelyan reviewed (approved):
Отлично!

> requestUpdate по аналогии с httpClient тоже хочется вынести в поле класса. При помощи какого-нибудь string builder, передавая разные части строки? Или не стоит так делать?

Разве что как приватную функцию. Попробуй в след. этапе, посмотрим как выглядит, насколько удобно
